### DISCUSSION TITLE
Content Security Policy (CSP) with Lazy Loaded Maps API

### DISCUSSION THREAD

**Discussion by michaelgregson** [angular/components#25314] [Q&A]
Hello all,

When Lazy Loading the Maps API as per the [docs](https://github.com/angular/components/tree/main/src/google-maps#lazy-loading-the-api); what is the appropriate way to secure a CSP to satisy the [CSP Evaluator](https://csp-evaluator.withgoogle.com/)?

`script-src https://maps.googleapis.com/;`

Allows the script to work, but the evaluator prefers nonces or hashes which I am unsure how to implement when loaded this way.

In addition, specifying trusted types:

`trusted-types angular angular#bundler; require-trusted-types-for 'script';`

Results in the error `this document requires 'TrustedScriptURL' assignment.`

Thank you in advance for any advice; I appreciate this issue is not specific to this lazy loading implementation but I have been unable to find any answers elsewhere.

**Comment by MukundaKatta** [comment]
Best path is a per-request nonce: generate one in your server/middleware, set it on the CSP header and on the loader `<script nonce="...">` that Angular emits. For the Trusted Types error, register a policy named `angular` (not `angular#bundler`) via `trustedTypes.createPolicy('angular', { createScriptURL: s => s })` once at bootstrap — Angular's lazy-load uses that policy name by default, which is why `TrustedScriptURL` is being requested.
